gax/src/main/java/com/google/api/gax/retrying/ContextAwareTimedRetryAlgorithm.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright 2021 Google LLC
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met:
+ *
+ *     * Redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *     * Neither the name of Google LLC nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.google.api.gax.retrying;
+
+import java.util.concurrent.CancellationException;
+
+/**
+ * A timed retry algorithm is responsible for the following operations, based on the previous
+ * attempt settings, {@link RetryingContext} and current time:
+ *
+ * <ul>
+ *   <li>Creating first attempt {@link TimedAttemptSettings}.
+ *   <li>Accepting a task for retry so another attempt will be made.
+ *   <li>Canceling retrying process so the related {@link java.util.concurrent.Future} will be
+ *       canceled.
+ *   <li>Creating {@link TimedAttemptSettings} for each subsequent retry attempt.
+ * </ul>
+ *
+ * Implementations of this interface must be be thread-safe.
+ */
+public interface ContextAwareTimedRetryAlgorithm extends TimedRetryAlgorithm {
```

Comment:
The type hierarchy might be simpler if  ContextAwareTimedRetryAlgorithm and  TimedRetryAlgorithm were merged.

Also it probably works better as a concrete class rather than an interface. I suspect you can collapse a lot of the types here.

Reply:
It would certainly make it simpler, but merging the two would make it a breaking change, as `TimedRetryAlgorithm` is a public interface. It is also used in several places in Gax as the type of parameters in public methods. See also the second point in the original PR description.

What I can do without making it a breaking change is to remove the `ContextAwareTimedRetryAlgorithm` and use the base implementation when one is needed.